Add possibility to provide a template to generate `host` and `advertisedHost` fields (#130)
strimzi/proposals: 084-templating-host-and-advertisedHost-fields.md

But in most cases, the (advertised) hostnames often follow the same pattern.
In the example above, you can see that the Ingress based listener uses the hostname pattern `my-cluster-kafka-<NODE-ID>.myingress.com`.
And the customized listener uses the advertised hostname pattern `broker-<NODE-ID>.strimzi.gateway.api.test`.

While there might be users using more random hostnames, for many Strimzi users, having an easy way to template the hostnames as suggested in this proposal would provide a huge simplification:
* Users will be able to scale-up without changing the configuration first
* The `Kafka` CR will be much smaller, simpler and easier to read
* The likelihood of typos and misconfigurations will be smaller

## Proposal

This proposal suggests adding two new fields to the `configuration` section of the listener:
* `advertisedHostTemplate`
* `hostTemplate`

These fields will be strings that will be used to configure templates that will be used to generate the hostnames.
The templates will support several different _variables_:
* `{nodeId}` will be replaced with the ID of the Kafka node to which the template is applied

> * `{nodeName}` will be replaced with the name of the Kafka node (name of the Kubernetes Pod used for this node) to which the template is applied

Reviewer: Generally a node can be thought of as having several 'names'. There's the pod name, but also host names (with and without qualification). It not really ambiguous here, because we're _configuring_ the host names. But if someone looked at the CR with the assumption that the `nodeName` was a host name then there's potential for confusion.

Another point to bear in mind, if the proposal is accepted then maybe we end up with other templates in CRs. Ideally we'd use a common set of variable names across all those places. In those other places `nodeName` might be more ambiguous.

For these reasons perhaps `nodePodName` would be better?

Author: This makes sense. I updated it to `nodePodName`.

> * `{nodeId}` will be replaced with the ID of the Kafka node to which the template is applied
> * `{nodePodName}` will be replaced with the name of the Kafka node (name of the Kubernetes Pod used for this node) to which the template is applied

Reviewer 2: ```suggestion
* The `{nodeId}` variable will be replaced with the ID of the Kafka node to which the template is applied.
* The `{nodePodName}` variable will be replaced with the Kubernetes pod name for the Kafka node where the template is applied.
```

Strimzi will take these template fields and replace the variables with the corresponding value for each Kafka node.
And the resulting value would be used for the (advertised) hostnames.

> The templates will be applied only to the per-broker values.
> The hostname for the bootstrap Ingress resource would still need to be specified directly and will never be generated from the template.
> (Advertised hostnames can be used only as per-broker configuration)

Author: I don't think the suggestion makes sense here ... in particular the `This is because` part.
* For the `host` field, the template is not used for the bootstrap `host` because the bootstrap `host` would use a different format/pattern. You cannot apply the template such as `broker-{nodeId}.myingress.com` to the bootstrap as there is no node ID. And you also have only one bootstrap `host`, so having a separate template makes no sense either.
* For the `advertisedHostname`, that is not supported at the bootstrap level because it is a configuration inside the Kafka configuration and can exist only per broker. (Whereas the `host` configuration is on the Kubernetes level)

Maybe we could use something like this?

```suggestion
The templates will only be applied to per-broker values. 
The hostname for the bootstrap Ingress resource must be specified manually and will not be generated from the template. 
The advertised hostnames are used only for per-broker configurations and cannot be configured for bootstrap.
```

> (Advertised hostnames can be used only as per-broker configuration)

Reviewer 2: works for me :+1:

### Handling conflicts

> It is possible that the user specifies both the `advertisedHostTemplate` / `hostTemplate` as well as the `advertisedHost` / `host` fields for one or more Kafka brokers.

Reviewer: Can we not use `oneOf`?

Author: I'm not sure OneOf would work given the fields are in very different places (+ one of them is inside an array). But I do not think it is needed. The fields fit well together if you need to override the template for some reason. We just need to establish the priority.

Author: @tombentley Is that ^^^ sufficient as an explanation here?

Reviewer: I think it's ok for this proposal. 

It would be nice if we could one day leverage [CRD validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules) for this kind of thing.

```

The YAML from the Gateway API blog post could be simplified as follows:

```yaml
listeners:
  - name: obiwan
    port: 9092
    tls: true
    type: cluster-ip
    configuration:
      brokerCertChainAndKey:
        certificate: tls.crt
        key: tls.key
        secretName: my-certificate
      advertisedHostTemplate: broker-{nodeId}.strimzi.gateway.api.test
```

## Affected projects

This proposal affects the Strimzi Cluster Operator only.

## Backwards compatibility

This proposal is fully backwards compatible.
The existing configurations with the (advertised) hostnames configured per-broker will continue to work without any change.

## Rejected alternatives

### Supporting additional variables

Originally, I considered supporting some additional variables in the template fields:
* `{clusterName}` will be replaced with the name of the Kafka cluster (name of the `Kafka` CR)
* `{namespace}` will be replaced with the namespace of the Kafka cluster (namespace where the `Kafka` CR exists)

But at the end I decided to not include them in the proposal.
These values do not differ between the different Kafka nodes and can be easily handled by the user.

> If the `Kafka` CR is generated from some kind template (e.g. Helm), the cluster name / namespace can be easily templated outside of Strimzi.

Reviewer 2: ```suggestion
If the `Kafka` CR is generated from some kind of template (e.g. Helm), the cluster name / namespace can be easily templated outside of Strimzi.
```
